### PR TITLE
for expression optimization

### PR DESCRIPTION
**Pull request by mrange**
(Moved from codeplex)
'for i in expr do body' optimization

Previously 'for i in expr do body' only had optimization
when the type of expr was an 1D-array.

'for i in expr do body' now has optimizations for when expr
is of type 'string' and 'List`1'.

These optimizations increases the performance of the for
expression but also reduces the number of objects created.

### CODE REVIEW

File: src/fsharp/FSharp.Core/prim-types.fsi [hunk 1]
```diff
@@ -1097,6 +1097,14 @@ namespace Microsoft.FSharp.Core
             //[<CompilerMessage("This function is for use by compiled F# code and should not be used directly", 1204, IsHidden=true)>]
             val inline GetString : source:string -> index:int -> char
 
+            /// <summary>A compiler intrinsic that gets character from a string</summary>
+            [<CompilerMessage("This function is for use by compiled F# code and should not be used directly", 1204, IsHidden=true)>]
+            val inline GetStringChar : source:string -> index:int -> char
```
Comment (mrange) [inline src/fsharp/FSharp.Core/prim-types.fsi]: I am looking for alternatives to add these methods to prim-types.fsi. The issue is that in the optimizer there is not so much support for find the correct overloads. In the type-checker it was possible to find the proper overload but migrating that code to the optimizer proved difficult.

**Comment by mrange** [comment]
Just wanted to get the PR started again.

File: src/fsharp/FSharp.Core/list.fs [hunk 2]
```diff
@@ -594,4 +594,4 @@ namespace Microsoft.FSharp.Collections
         let truncate count list = Microsoft.FSharp.Primitives.Basics.List.truncate count list
 
         [<CompiledName("Unfold")>]
-        let unfold<'T,'State> (f:'State -> ('T*'State) option) (s:'State) = Microsoft.FSharp.Primitives.Basics.List.unfold f s
+        let unfold<'T,'State> (f:'State -> ('T*'State) option) (s:'State) = Microsoft.FSharp.Primitives.Basics.List.unfold f s
```
Comment (enricosada) [inline src/fsharp/FSharp.Core/list.fs]: useless whitespace change. Also newline at end of file is good

File: src/fsharp/env.fs [hunk 3]
```diff
@@ -791,7 +796,7 @@ let mkTcGlobals (compilingFslib,sysCcu,ilg,fslibCcu,directoryToResolveRelativePa
                             fslib_MFQueryRunExtensionsHighPriority_nleref 
 
                             fslib_MFSeqModule_nleref    
-                            fslib_MFListModule_nleref   
+                            fslib_MFListModule_nleref
```
Comment (enricosada) [inline src/fsharp/env.fs]: whitespace

File: tests/fsharpqa/Source/Optimizations/ForLoop/ForEachOnString01.fs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+// multiline body        
+let test9() = 
+    for i in ( 
+                "1234"
+                |> String.map (fun x ->
+                    char ((int x) + 1))
+                ) do
+        let tmp = System.String.Format("{0} foo", i)
+        printfn "%O" tmp
```
Comment (enricosada) [inline tests/fsharpqa/Source/Optimizations/ForLoop/ForEachOnString01.fs]: newline at eof

File: tests/fsharp/core/measures/test.fsx [hunk 5]
```diff
@@ -15,12 +15,13 @@ let test s b = if b then () else report_failure(s)
...
-    let cultureString = argv.[2] in 
-    let culture = new System.Globalization.CultureInfo(cultureString) in 
+  if argv.Length > 2 && argv.[1] = "--culture" then
+    let cultureString = argv.[2]
+    let culture = new System.Globalization.CultureInfo(cultureString)
     stdout.WriteLine ("Running under culture "+culture.ToString()+"...");
     System.Threading.Thread.CurrentThread.CurrentCulture <-  culture
-  end 
+  else
+    System.Threading.Thread.CurrentThread.CurrentCulture <-  System.Globalization.CultureInfo.InvariantCulture
```
Comment (enricosada) [inline tests/fsharp/core/measures/test.fsx]: this fix a failing test not related to pr?

**Comment by dsyme** [comment]
I think there's a problem with this PR when using FSharp.Core 4.3.0.0 or 4.3.1.0

The F# 4.0 compiler must be able to successfully work with FSHarp.Core 4.3.0.0 and 4.3.1.0.  The GetStringChar and GetStringLength intrinsics won't be available in those libraries (TryDeref on the ValRef will return None).  For those cases, we need to make sure the optimization doesn't apply

@KevinRansom  @latkin - This reminds me that we really need to beef up our testing for this backward-library-targeting case.  I wonder if the "single file test and run" in the "fsharp" suite needs to be beefed up with a compilation against both 4.3.0.0 and 4.3.1.0, with a #define to cover the API differences between the libraries.

**Comment by latkin** [comment]
Yes, when reviewing this on Codeplex, it got to a good state for F# 4.0-only.  But design causes the 4.0 compiler to error out when it encounters one of these loops and is targeting 3.1.0 or earlier. This needs to be resolved.

@dsyme yes, this back-compat stuff is on my backlog. I have some ideas, just need to find time.

**Comment by dsyme** [comment]
@mrange - I think you just need to add ``mkstringchar_vref`` to TcGlobals and check ``mkstringchar_vref.CanDeref`` as part of the case guard here: https://github.com/Microsoft/visualfsharp/pull/219/files#diff-b6f3f390e9807a7e9d93964104d74363R7819

Likewise for any other cases using new constructs

**Comment by mrange** [comment]
Hi.

Reworked the code so that it no longer relies on new functions being added to IntrinsicFunctions.

That should make the change backwards comptatible with FSharp.Core 4.3.0.0 and 4.3.1.0.

It's getting a bit late here so hopefully I haven't lost latkins portable fixes along the way.

**Comment by mrange** [comment]
AppVeyor seemed to break with: 
Setting up Visual Studio for debugging extensions. This one-time operation may take a minute or more.
 1575  C:\Program Files (x86)\Microsoft Visual Studio 12.0\Common7\IDE\devenv.exe /RootSuffix FSharp /ResetSettings General.vssettings /Embedding /Command File.Exit 
 1576log4net : error XmlHierarchyConfigurator: Could not create Appender [RollingFileAppender] of type [log4net.Appender.RollingFileAppender]. Reported error follows. [C:\projects\visualfsharp-radou\vsintegration\src\deployment\EnableOpenSource\EnableOpenSource.csproj]
 1577  System.IO.FileNotFoundException: Could not load file or assembly 'AWSToolkit.Util' or one of its dependencies. The system cannot find the file specified.
 1578  File name: 'AWSToolkit.Util'

**Comment by mrange** [comment]
Tried a quick smoke test on 4.3.0.0 and 4.3.1.0 and it seems to work now.

**Comment by mrange** [comment]
Btw, I don't expect this to be merged until after 4.0 is released as I realize this is somewhat risky addition. Yet I want to get it to a mergable state.

File: src/fsharp/tastops.fs [hunk 6]
```diff
@@ -7779,34 +7796,92 @@ let (|RangeInt32Step|_|) g expr =
...
-              TryFinally (WhileLoopForCompiledForEachExpr (_guardExpr, Let (elemVar,_currentExpr,_,bodyExpr), m), _cleanupExpr))) -> 
+let (|ExtractTypeOfExpr|_|) g expr = Some (tyOfExpr g expr)
 
-        let spForLoop = match spBind with SequencePointAtBinding(spStart) -> SequencePointAtForLoop(spStart) |  _ -> NoSequencePointAtForLoop 
+type OptimizeForExpressionOptions = OptimizeIntRangesOnly | OptimizeAllForExpressions
 
-        Some(spForLoop,elemVar,startExpr,step,finishExpr,bodyExpr,m)
-    | _ -> 
-        None
+let DetectAndOptimizeForExpression g option expr =
```
Comment (mrange) [inline src/fsharp/tastops.fs]: The exciting stuff

**Comment by latkin** [comment]
working on validating/merging this guy now

**Comment by latkin** [comment]
Applied to the OOB branch. Thanks for being patient on this one, it's great to finally have this merged!  I added cross-version tests and verified things are good in that regard.

**Pull request closed without merging**